plugin/pkg/shoot/dns/admission.go

```diff
@@ -149,56 +151,59 @@ func (d *DNS) Admit(a admission.Attributes, o admission.ObjectInterfaces) error
 // assignDefaultDomainIfNeeded generates a domain <shoot-name>.<project-name>.<default-domain>
 // and sets it in the shoot resource in the `spec.dns.domain` field.
 // If for any reason no domain can be generated, no domain is assigned to the Shoot.
-func assignDefaultDomainIfNeeded(shoot *garden.Shoot, projectLister gardenlisters.ProjectLister, secretLister kubecorev1listers.SecretLister) {
+func assignDefaultDomainIfNeeded(shoot *garden.Shoot, projectLister gardenlisters.ProjectLister, secretLister kubecorev1listers.SecretLister) error {
 	project, err := admissionutils.GetProject(shoot.Namespace, projectLister)
 	if err != nil {
-		return
+		return apierrors.NewInternalError(err)
 	}
 
 	var domainSecrets []*corev1.Secret
 	deprecatedSelector, err := labels.Parse(fmt.Sprintf("%s=%s", v1alpha1constants.DeprecatedGardenRole, common.GardenRoleDefaultDomain))
 	if err != nil {
-		return
+		return apierrors.NewInternalError(err)
 	}
 	secrets, err := secretLister.Secrets(v1alpha1constants.GardenNamespace).List(deprecatedSelector)
 	if err != nil {
-		return
+		return apierrors.NewInternalError(err)
 	}
 	domainSecrets = append(domainSecrets, secrets...)
 
 	selector, err := labels.Parse(fmt.Sprintf("%s=%s", v1alpha1constants.GardenRole, common.GardenRoleDefaultDomain))
 	if err != nil {
-		return
+		return apierrors.NewInternalError(err)
 	}
 	secrets, err = secretLister.Secrets(v1alpha1constants.GardenNamespace).List(selector)
 	if err != nil {
-		return
+		return apierrors.NewInternalError(err)
 	}
 	domainSecrets = append(domainSecrets, secrets...)
 
 	if len(domainSecrets) == 0 {
-		return
+		return nil
 	}
 
 	var shootDomain *string
 	if shoot.Spec.DNS != nil {
 		shootDomain = shoot.Spec.DNS.Domain
 	}
 
-	if shootDomain != nil {
-		return
-	}
-
 	for _, secret := range domainSecrets {
 		_, domain, _, _, err := common.GetDomainInfoFromAnnotations(secret.Annotations)
 		if err != nil {
-			return
+			return apierrors.NewInternalError(err)
 		}
 
-		if shootDomain != nil && strings.HasSuffix(*shootDomain, domain) {
+		if shootDomain != nil && strings.HasSuffix(*shootDomain, "."+domain) {
 			// Shoot already specifies a default domain, set providers to nil
 			shoot.Spec.DNS.Providers = nil
-			return
+
+			// Check that the specified domain matches the pattern for default domains, especially in order
+			// to prevent shoots from "stealing" domain names for shoots in other projects
+			if *shootDomain != fmt.Sprintf("%s.%s.%s", shoot.Name, project.Name, domain) {
```

Comment:
With this condition we would only allow users to specify the subdomain pattern `<shoot>.<project>.` if the default domain is used.  
This would not only forbid "stealing" default domains from other project but would also forbid using different domain names in your own project subdomain (for example, using `foo.<project>.<default-domain>` for a Shoot named `bar` would not be possible).

If the PR is only about forbidding users to "steal" domains from other projects, I think this check should be improved to allow the describe scenario.
WDYT?

Reply:
Well, I don't see of any scenario where doing something like this is useful. If somebody wants to play with the domain names then he can use his own domain instead of our default domain. If our default domain is used we are dictating the scheme. Allowing stuff like this will only enable the user to harm itself (he can never create a shoot named `foo` using the default domain), so, I don't think it makes sense to maintain code for such a scenario.

Reply:
Agreed, you are right. Then I think its fine 👍